### PR TITLE
Last-Element Padding for Partial Offset Tensors

### PR DESCRIPTION
**Pull request by allwu**
Summary:
Add last-element padding support in NNPI backend for two torch_glow nodes:
- EmbeddingBagNode
- EmbeddingBagByteRowwiseOffsetsNode
The padding is done in host, with helper functions similar to ```BackendUtils::allowsPartialInput```.

Differential Revision: D21731699

### CODE REVIEW

**Comment by jfix71** [comment]
CC: @jsubag

File: lib/Backends/NNPI/NNPIResource.cpp [hunk 1]
```diff
@@ -544,13 +545,24 @@ bool NNPIResource::updateHostResourceFromTensor(Tensor *t, bool partialTensor) {
...
   if (partialData && !partialTensor) {
-    memset(reinterpret_cast<uint8_t *>(hostPtr_) + unpaddedSize, 0,
-           paddedSize - unpaddedSize);
+    if (!requiresNZPadding) {
+      memset(reinterpret_cast<uint8_t *>(hostPtr_) + unpaddedSize, 0,
+             paddedSize - unpaddedSize);
+    } else {
+      LOG_AND_RETURN_IF(
+          ERROR, !downcastInt64,
+          "Non-zero padding is not supported for none-index types", false);
```
Comment (jsubag) [inline lib/Backends/NNPI/NNPIResource.cpp:554]: typo: shouldn't it be "non-index"

File: lib/Backends/NNPI/NNPIResource.cpp [hunk 2]
```diff
@@ -544,13 +545,24 @@ bool NNPIResource::updateHostResourceFromTensor(Tensor *t, bool partialTensor) {
...
   // Pad with zeros if needed.
   if (partialData && !partialTensor) {
-    memset(reinterpret_cast<uint8_t *>(hostPtr_) + unpaddedSize, 0,
-           paddedSize - unpaddedSize);
+    if (!requiresNZPadding) {
+      memset(reinterpret_cast<uint8_t *>(hostPtr_) + unpaddedSize, 0,
+             paddedSize - unpaddedSize);
+    } else {
+      LOG_AND_RETURN_IF(
+          ERROR, !downcastInt64,
```
Comment (jsubag) [inline lib/Backends/NNPI/NNPIResource.cpp:553]: You may want to amend this condition. There's nothing preventing padding for any data type and in the future you may want to change the model to already hold indices in 32bit and save the conversion from 64bit.

File: lib/Backends/NNPI/NNPIResource.cpp [hunk 3]
```diff
@@ -544,13 +545,24 @@ bool NNPIResource::updateHostResourceFromTensor(Tensor *t, bool partialTensor) {
...
+          "Non-zero padding is not supported for none-index types", false);
+      int numElements = unpaddedSize / sizeof(int32_t);
+      int totalElements = paddedSize / sizeof(int32_t);
+      std::fill(reinterpret_cast<int32_t *>(hostPtr_) + numElements,
+                reinterpret_cast<int32_t *>(hostPtr_) + totalElements,
+                reinterpret_cast<const int32_t *>(hostPtr_)[numElements - 1]);
+    }
+  } else if (partialData) {
+    // Update partial size
+    partialSize_ = unpaddedSize;
```
Comment (jsubag) [inline lib/Backends/NNPI/NNPIResource.cpp:563]: partialSize_ needs to be updated regardless (i.e. in this version 0 if not padding).
Later parts in the code are relying on it.
So not updating will carry over padded size from one inference to the next.

**Comment by jsubag** [comment]
CC: @omromano

**Comment by allwu** [comment]
@jsubag thanks for your comments! I have addressed them, please let me know if you have any other comments.

**Comment by jsubag** [comment]
LGTM.

**Pull request closed without merging**